File: client2/src/game/systems/JumpReset.ts
import { System } from '../../ecs/System'
import { SystemFlag } from '../flags'
import { NeverAspect } from '../../ecs/Aspect'
import { Movement } from '../components/Movement'
import { check, hasComponent, isNotBodyType } from '../utils/collision'
import { Player } from '../components/Player'

export class JumpResetSystem extends System {
  static _id = 'JumpReset'
  _id = JumpResetSystem._id

  static _flag = SystemFlag.JumpReset
  _flag = JumpResetSystem._flag

  static _aspect = new NeverAspect()
  _aspect = JumpResetSystem._aspect

  beginContact = (a: Fixture, b: Fixture, contact: Contact) => {
    const result = check(a, b, [hasComponent(Player), isNotBodyType('dynamic')])

    if (!result) {
      return
    }

    const [player] = result
    const [nx, ny] = contact.getNormal()

    if (nx == -1 || nx == 1 || ny < 0) {
      const movement = player.as(Movement)

      if (movement) {
        movement.jumpCount = 0
      }
    }
  }
}

File: archive/client/src/game/systems/jumpreset.lua
local Aspect = require 'src.ecs.aspect'
local System = require 'src.ecs.system'
local Fixture = require 'src.game.components.fixture'
local Movement = require 'src.game.components.movement'
local Player = require 'src.game.components.movement'
local collision = require 'src.game.utils.collision'

local JumpReset = System:new('jumpreset', Aspect.never())

function JumpReset:beginContact(a, b, contact)
  local nx, ny = contact:getNormal()

  local entities =
    collision.check(
    a,
    b,
    {collision.hasComponent(Player), collision.isNotBodyType('dynamic')}
  )

  if entities and (nx == -1 or nx == 1 or ny < 0) then
    entities[1]:as(Movement).jumpCount = 0
  end
end

return JumpReset
